File: Go/Anagrams/438. Find All Anagrams in a String.go
package main

import (
	"fmt"
)

func findAnagrams(s string, p string) []int {

	Ans := []int{}
	if len(s) < len(p) {
		return Ans
	}
	m1 := make([]int, 26)
	m2 := make([]int, 26)
	for i := range p {
		m1[p[i]-'a']++
		m2[s[i]-'a']++
	}
	var matches int
	for i := range m1 {
		if m1[i] == m2[i] {
			matches++
		}
	}
	for i := len(p); i <= len(s); i++ {
		if matches == 26 {
			Ans = append(Ans, i-len(p)+1)
		}
		if i == len(s) {
			break
		}
		l := s[i-len(p)] - 'a'
		r := s[i] - 'a'
		m2[r]++
		if m2[r] == m1[r] {
			matches++
		} else if m2[r] == m1[r]+1 {
			matches--
		}
		m2[l]--
		if m2[l] == m1[l] {
			matches++
		} else if m2[l] == m1[l]-1 {
			matches--

		}
	}
	return Ans
}
func main() {
	fmt.Println(findAnagrams("baa", "aa"))
	fmt.Println(findAnagrams("abab", "ab"))
}

File: Leetcode/438. Find All Anagrams in a String.py
class Solution(object):
    def getID(self, s1):
        id=[0]*26
        for x in s1:
            id[ord(x)-97]+=1
        return id
    def findAnagrams(self, s, p):
        if len(p)>len(s): return []
        pID=self.getID(p)
        Wid=self.getID(s[0:len(p)])
        matches=0
        for i in range(0,26):
            if pID[i]==Wid[i]: matches+=1
        Ans=[]
        if matches==26: Ans.append(0)
        ndx=len(p)
        while(ndx<len(s)):
            Wid[ord(s[ndx])-97]+=1
            if Wid[ord(s[ndx])-97]==pID[ord(s[ndx])-97]:    matches+=1
            elif Wid[ord(s[ndx])-97]==pID[ord(s[ndx])-97]+1: matches-=1
            Wid[ord(s[ndx-len(p)])-97]-=1
            if Wid[ord(s[ndx-len(p)])-97]==pID[ord(s[ndx-len(p)])-97]:  matches+=1
            elif Wid[ord(s[ndx-len(p)])-97]==pID[ord(s[ndx-len(p)])-97]-1: matches-=1
            if matches==26:  Ans.append(ndx-len(p)+1)
            ndx+=1
        return Ans
if __name__ == "__main__":
    obj=Solution()
    Op=obj.findAnagrams(s = "bpaa", p = "aa")
    print(Op)